rule runHtSeqCount:
	input:
		SAMFILE="{replicat}/{condition}/AlignementStarWithChimeric/{type}/{sample}_{type}_Aligned.out.sam"
	params:
		GTF=config["ANNOTATION"]["PATH"],
		tmpSAM="{replicat}/{condition}/AlignementStarWithChimeric/{type}/{sample}_{type}_AlignedWithoutSoftClipped.sam"
	output:
		HtSeqCountOutput="{replicat}/{condition}/Global_results/{sample}{condition}_{type}_htSeqCount.txt"
	run:
		shell("samtools view -h -F4 {input.SAMFILE} > {params.tmpSAM} ")
		shell("htseq-count --stranded=yes --type=gene --idattr=gene_id --nonunique=all --secondary-alignments=ignore --quiet {params.tmpSAM} {params.GTF} | awk '$1 !~ /__/'  > {output.HtSeqCountOutput}"),
		shell(" rm {params.tmpSAM}"),
